src/react/utils.js

```diff
@@ -58,6 +60,95 @@ export function valueIsClassComponent(realm: Realm, value: Value) {
   return false;
 }
 
+// a nested object of a React Element should be hoisted where all its properties are known
+// at evaluation time to be "static" or "pure". This means function values in scope, strings, numbers
+// and deeply nested object/arrays that share the same heuristics
+function canHoistObject(realm: Realm, object: ObjectValue, hoistedScope?: Generator): boolean {
+  if (isReactElement(object)) {
+    return canHoistReactElement(realm, object);
+  }
+  for (let [propName] of object.properties) {
+    let prop = Get(realm, object, propName);
+    if (prop instanceof AbstractValue) {
+      return false;
+    } else if (prop instanceof FunctionValue && prop.parent !== hoistedScope) {
+      return false;
+    } else if (prop instanceof ArrayValue) {
+      let isHoistable = canHoistArray(realm, prop, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    } else if (prop instanceof ObjectValue) {
+      let isHoistable = canHoistObject(realm, prop, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    }
+  }
+  return true;
```

Comment:
So this is not strictly safe because these objects are mutable and any mutation would be shared.

We probably should document the rationale for why we think that this is safe so that we can compare against that rationale to see if this is the correct implementation.

The rationale used in the Babel transform might not apply here because that might use things like object literals as a heuristic where this doesn't.

We also have another thing that is not strictly safe where if two JSX elements have referential equality, we always bailout regardless of sCU. The rationale is that most JSX elements are created in render functions and don't live in state with mutations.

I believe that it probably is safe given the rationale above and the fact that this optimization only applies in the particular optimization we already do. But I'm not 100% sure this is safe yet.